astropy/utils/masked/function_helpers.py

```diff
@@ -1055,7 +1058,11 @@ def array2string(
 
 @dispatched_function
 def array_str(a, max_line_width=None, precision=None, suppress_small=None):
-    # Override to avoid special treatment of array scalars.
+    # Override to change special treatment of array scalars, since the numpy
+    # code turns the masked array scalar into a regular array scalar.
+    if a.shape == () and a.dtype.names is None:
+        return MaskedFormat.maybe_mask_string(np.array_str(a.unmasked), a.mask)
```

Comment:
that would follow through to here.

Reply:
I now use `MaskedFormat` explicitly, with `np.array_str` as the format function. That seems cleaner.

Eventually, I'd like to give the option of using different ways to indicate masked - in particular, with strikethrough! (but it is tricky since not all terminals support the utf-8 way of doing it; see line 965)